contracts/multisig-prover/src/contract.rs

```diff
@@ -1,44 +1,479 @@
+            }
+        };
+
+        // keep track of the batch id to use during submessage reply
+        REPLY_BATCH.save(deps.storage, &command_batch.id)?;
+
+        let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+            key_id,
+            msg: command_batch.msg_to_sign(),
+        };
+
+        let wasm_msg = wasm_execute(config.multisig, &start_sig_msg, vec![])?;
+
+        Ok(Response::new()
+            .add_submessage(SubMsg::reply_on_success(wasm_msg, START_MULTISIG_REPLY_ID)))
+    }
+
+    pub fn rotate_snapshot(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        pub_keys: HashMap<String, HexBinary>,
+        key_id: String,
+    ) -> Result<Response, ContractError> {
+        KEY_ID.save(deps.storage, &key_id)?;
+
+        let config = CONFIG.load(deps.storage)?;
+        if config.admin != info.sender {
+            return Err(ContractError::Unauthorized {});
+        }
+
+        let snapshot = snapshot(deps.querier, env.block, &config)?;
+
+        let keygen_msg = WasmMsg::Execute {
+            contract_addr: config.multisig.into(),
+            msg: to_binary(&multisig::msg::ExecuteMsg::KeyGen {
+                key_id: key_id.clone(),
+                snapshot: snapshot.clone(),
+                pub_keys: pub_keys.clone(),
+            })?,
+            funds: vec![],
+        };
+
+        let event = Event::SnapshotRotated {
+            key_id,
+            snapshot,
+            pub_keys,
+        };
+
+        Ok(Response::new()
+            .add_message(keygen_msg)
+            .add_event(event.into()))
+    }
+
+    fn snapshot(
+        querier: QuerierWrapper,
+        block: BlockInfo,
+        config: &Config,
+    ) -> Result<Snapshot, ContractError> {
+        let query_msg = service_registry::msg::QueryMsg::GetActiveWorkers {
+            service_name: config.service_name.clone(),
+            chain_name: config.chain_name.to_string(),
+        };
+
+        let active_workers: Vec<Worker> = querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.service_registry.to_string(),
+            msg: to_binary(&query_msg)?,
+        }))?;
+
+        let participants = active_workers
+            .into_iter()
+            .map(Worker::try_into)
+            .collect::<Result<Vec<Participant>, service_registry::ContractError>>()
+            .map_err(
+                |err: service_registry::ContractError| ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                },
+            )?
+            .try_into()
+            .map_err(|err: axelar_wasm_std::nonempty::Error| {
+                ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                }
+            })?;
+
+        Ok(Snapshot::new(
+            block
+                .time
+                .try_into()
+                .expect("violated invariant: block time cannot be invalid"),
```

Comment:
```suggestion
                .expect("violated invariant: block time cannot be zero"),
```

Reply:
it used to be `zero` before, but I changed it because from the contract perspective `axelar_wasm_std` is an external library and it shouldn't assume that the error resulting from `try_into` will always be because it's zero

Reply:
what else could it be?